posix: Implement pthread_equal function

jaiiarora opened the pull request:
Patch implements pthread_equal function.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/59942

jaiiarora committed:
posix: Changes pthread_equal to regular type function

Patch changes pthread_equal function from static inline to regular
function type

Fix for issue#59942

Signed-off-by: Jai Arora <infolinesoni@gmail.com>

cfriedt reviewed (commented):
@jaiiarora - Looks good, but can you please also add a second commit with tests in `tests/posix/common/src/pthread.c`?

Also, can you update your commit message? Technically, `pthread_equal()` was implemented, but your change converts it from a static inline function to a regular function.

jaiiarora commented:
Okay, I will work on it

jaiiarora commented:
add required changes @cfriedt

jaiiarora commented:
Please let me know if anything  is missing @cfriedt. I will work on it

cfriedt commented on tests/posix/common/src/pthread.c:
```diff
@@ -30,6 +30,7 @@
 K_THREAD_STACK_ARRAY_DEFINE(stack_e, N_THR_E, STACKS);
 K_THREAD_STACK_ARRAY_DEFINE(stack_t, N_THR_T, STACKS);
 K_THREAD_STACK_ARRAY_DEFINE(stack_1, 1, 32);
+K_THREAD_STACK_DEFINE(stack_p, STACKS);
```
This can be removed.

jaiiarora replied:
done

cfriedt commented on tests/posix/common/src/pthread.c:
```diff
@@ -771,3 +772,58 @@ ZTEST(posix_apis, test_posix_pthread_barrier)
+	struct sched_param schedparam;
+	int schedpolicy = SCHED_FIFO;
+
+	/* iniitialize pthread attributes */
+	ret = pthread_attr_init(&attr);
+	zassert_equal(ret, 0, "pthread_attr_init failed");
+
+	schedparam.sched_priority = 1;
+	pthread_attr_setstack(&attr, &stack_p, STACKS);
+	zassert_equal(ret, 0, "pthread_attr_setstack failed");
+	pthread_attr_setschedpolicy(&attr, schedpolicy);
+	zassert_equal(ret, 0, "pthread_attr_setschedpolicy failed");
+	pthread_attr_setschedparam(&attr, &schedparam);
+	zassert_equal(ret, 0, "pthread_attr_setschedparam failed");
+
+	/* Create thread */
+	ret = pthread_create(&thread_id, &attr, thread_func, NULL);
+	zassert_equal(ret, 0, "pthread_create failed");
+
+	ret = pthread_join(thread_id, NULL);
+	zassert_equal(ret, 0, "pthread_join failed");
+
+	printk("thread_id %u\n", thread_id);
+	printk("thread_glb %u\n", thread_glb);
+
+	/* both should have same value */
+	if (!pthread_equal(thread_id, thread_glb))
+		zassert_true(false, "pthread_equal failed");
+
+	printk("child thread_id is %u\n", thread_id);
+	printk("parent thread_id is %u\n", pthread_self());
+
+	/* both should have different id */
+	if (pthread_equal(thread_id, pthread_self()))
+		zassert_true(false, "pthread_equal failed");
+
+	ret = pthread_attr_destroy(&attr);
+	zassert_equal(ret, 0, "pthread_attr_destroy failed");
+
+	printk("pthread_equal test ok\n");
```
```suggestion
ZTEST(posix_apis, test_pthread_equal)
{
	zassert_true(pthread_equal(pthread_self(), pthread_self()));
	zassert_false(pthread_equal(pthread_self(), (pthread_t)4242);
```

jaiiarora replied:
added requested changes

jaiiarora committed:
posix: Adds test case for pthread_equal functions

Adds test case for pthread_equal
Adds test coverage for same and different thread id

Signed-off-by: Jai Arora <infolinesoni@gmail.com>

jaiiarora committed:
docs: update posix docs with pthread_equal

add pthread_equal to supported api list

Signed-off-by: Jai Arora <infolinesoni@gmail.com>

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -848,4 +848,9 @@ static int posix_thread_pool_init(void)
 
+int pthread_equal(pthread_t pt1, pthread_t pt2)
+{
+	return (pt1 == pt2);
```
nit: no brackets needed here, but I'm not going to stop this from going in now. If there is another review that requires another spin, please fix this too.

cfriedt reviewed (approved):
LGTM - thanks @jaiiarora 👍
